plug.vim

```
@@ -968,6 +973,377 @@ while 1 " Without TCO, Vim stack is bound to explode
+    self.timeout = timeout
+    self.ntries = ntries
+
+    self.clean = clean
+    if cb:
+      self.callback = cb
+    else:
+      def dummy(msg): pass
+      self.callback = dummy
+
+  def attempt_cmd(self):
+    """ Looping call to attempt cmd ntries with a timeout. """
+    attempt = 0
+    finished = False
+    limit = self.timeout
+
+    while not finished:
+      try:
+        attempt += 1
+        result = self.timeout_cmd()
+        finished = True
+      except CmdTimedOut:
+        if attempt != self.ntries:
+          for count in range(3, 0, -1):
+            msg = 'Timeout. Will retry in {} second{} ...'.format(
+                count, 's' if count != 1 else '')
+            self.callback([msg])
+            time.sleep(1)
+          self.timeout += limit
+          self.callback(['Retrying ...'])
+        else:
+          raise
+
+    return result
+
+  def timeout_cmd(self):
+    """ Execute a cmd and returns output. Output is list of lines. BLOCKING.
+    Raises CmdFailed   -> return code for Popen isn't 0
+    Raises CmdTimedOut -> command exceeded timeout without new output
+    """
+    pcmd = None
+    first_line = True
+    try:
+      tfile = tempfile.NamedTemporaryFile()
+      pcmd = subprocess.Popen(self.cmd, cwd=self.cmd_dir, stdout=tfile,
+          stderr=subprocess.STDOUT, shell=True)
+      while pcmd.poll() == None:
+        # Yield this thread
+        time.sleep(0)
+
+        if first_line or random.random() < G_LOG_PROB:
+          first_line = False
+          line = nonblock_read(tfile.name)
+          if line and self.callback:
+            self.callback([line])
+
+        time_diff = time.time() - os.path.getmtime(tfile.name)
+        if time_diff > self.timeout:
+          raise CmdTimedOut(['Timeout!'])
+
+      tfile.seek(0)
+      result = [line.rstrip() for line in tfile]
+
+      if pcmd.returncode != 0:
+        msg = ['CmdFailed:', self.cmd]
+        msg.extend(result)
+        raise CmdFailed(msg)
+    except:
```

I am aware of this, I'm kind of abusing broad `except:` lines to act as finally clauses that only execute on exceptional behaviour. Note, I am not suppressing the exceptions, I raise at the end. Just cleaning things up like my subprocesses, which should be cleaned up regardless what throws.

I wonder if python devs do this or use a boolean in a finally clause.